source/getting-started/advanced-tutorials/fpga-ip-core/index.md

```diff
@@ -0,0 +1,611 @@
+
+```{image} images/vivado_edit_ip3.png
+:width: 500px
+```
+
+Now, read requests at offset "2" (i.e., for the second slave register `slv_reg2`) will return our adder module output, not the actual contents of `slv_reg2`.
+
+It is interesting to think about *write* operations to `slv_reg2` from the C code.
+The template Verilog will still update the value stored in `slv_reg2`, but this has no effect.
+To the user, `slv_reg2` has become a "read-only" register, as desired!
+
+
+### 7. Re-Package IP Core
+
+Following the [steps from above](#editing-ip-cores), re-package `my_custom_adder` IP core and return to the top-level FPGA block design.
+
+1. Ensure all Verilog source files are saved
+2. Go back to the `Package IP` pane
+3. Notice that the `File Groups` step is not a green checkbox
+4. Click the `File Groups` step and click `Merge changes from File Groups Wizard`
+5. Now, all the steps are green except the last
+6. Select the `Review and Package` step
+7. Click `Re-Package IP`
+8. Click `Yes` to close the temporary project
+
+### 8. Upgrade the IP
+
+Follow the [above steps](#editing-ip-cores) to upgrade the IP status.
+
+### 9. Generate the Bitstream
+
+Finally, click the `Generate Bitstream` as usual to create the bitstream file.
+
+### 10. Export Hardware
+
+After successful bitstream generation, re-export the hardware from Vivado, like was done in the [building and running firmware guide](/firmware/xilinx-tools/building-and-running-firmware). You should see a pop-up in the SDK after re-exporting the hardware. Select `Yes` in the pop-up to confirm the hardware export from Vivado.
+
+**Whew! The new custom IP core is complete!**
+
+## Access IP Core from C Code
+
+Per the above section and discussion, you should now be aware that the access to the IP core comes from the slave registers which the AXI4-Lite interface reads and writes. Let's understand our new IP core from the prospective of the C code.
+
+### Memory-Mapped FPGA Registers
+
+The core "magic" of the AXI4-Lite interface to the FPGA is the hardware support in the PS for memory operations.
+When the C code reads/writes to memory addresses near the IP core base address, this causes an AXI read/write transaction to occur.
+The template Verilog from the IP Wizard implements the slave side for the transaction response.
+
+Let's read/write from/to a slave register in our new IP core from C code:
+
+```C
+// Base address of our new IP core
+// 
+// This is simply a pointer with a manually assigned address per Vivado,
+// which might seem odd since we never do this in typical programming
+//
+// Recall this value came from the Address Editor pane in Vivado,
+// and also gets exported into the xparameters.h file
+volatile uint32_t *base_addr = (volatile uint32_t *) 0x43DB0000;
+
+// Define pointer to slv_reg3 (it was unused in the FPGA design)
+volatile uint32_t *slv_reg3 = &base_addr[3];
```

Comment:
Why aren't we showing how to get the base address using the define in xparameters.h?

Reply:
Eh, it should explain this...

It is somewhat complicated since the `xparameters.h` file is only generated from the SDK side when the `bsp` projects are created from the `hardware_wrapper` project, which itself is created via the `hdf` file which is exported from Vivado.......

I'll just add that:

> Once the SDK environment is set up, all IP core base addresses in the FPGA are available as auto-generated defines in the `xparameters.h` file.

Reply:
Done